README.md
# Domain Port Check

[![Latest Version on Packagist](https://img.shields.io/packagist/v/gemz/port.svg?style=flat-square)](https://packagist.org/packages/gemz/port)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/gemzio/port/run-tests?label=tests)](https://github.com/gemzio/port/actions?query=workflow%3Arun-tests+branch%3Amaster)
[![Quality Score](https://img.shields.io/scrutinizer/g/gemzio/port.svg?style=flat-square)](https://scrutinizer-ci.com/g/gemzio/port)
[![Total Downloads](https://img.shields.io/packagist/dt/gemz/port.svg?style=flat-square)](https://packagist.org/packages/gemz/port)

Check ports with protocol. This package uses under the hood [ReactPHP Promises](https://github.com/reactphp/promise).

## Installation

You can install the package via composer:

```bash
composer require gemz/port
```

## Usage

``` php
use \Gemz\Port\Port;

$checks = new Port('gemz.io');
// or
$checks = Port::for('gemz.io');

// check all default ports on tcp
$checks = Port::for('gemz.io')->check();

// check specific ports on tcp
$checks = Port::for('gemz.io')->check(80, 8080, 443, 22, 3306, 9000, 9001);

// check only specific ports on tcp
$checks = Port::for('gemz.io')->useTcp()->check(80, 8080);

// check only specific ports on udp
$checks = Port::for('gemz.io')->useUdp()->check(110, 140);

// check only specific ports on tls
$checks = Port::for('gemz.io')->useTls()->check(443);

// check only specific ports on ssl
$checks = Port::for('gemz.io')->useSsl()->check(443);

// check with array for specific port => protocol checks
// if global setting will be ignored
$checks = Port::for('gemz.io')->useTcp()->check([80 => 'tcp', 2525 => 'udp', 443 => 'tls']);

// check with array for specific port 
$checks = Port::for('gemz.io')->useTcp()->check([80, 2525, 443]);

// set timeout, default is 0.5s
$checks = Port::for('gemz.io')->setTimeout(0.4)->check(80);

// get supported protocols
$protocols = Port::for('gemz.io')->getProtocols();

// get default ports
$ports = Port::for('gemz.io')->getDefaultPorts();

// get domain
$protocols = Port::for('gemz.io')->getDomain();

```

### Testing

``` bash
composer test
composer test-coverage
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email stefan@sriehl.com instead of using the issue tracker.

## Credits

- [Stefan Riehl](https://github.com/stefanriehl)
- [All Contributors](../../contributors)

## Support us

Gemz.io is maintained by [Stefan Riehl](https://github.com/stefanriehl). You'll find all open source
projects on [Gemz.io github](https://github.com/gemzio).

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `port` will be documented in this file

## 1.0.0 - 2020-04-19

- initial release


tests/PortTest.php
<?php

namespace Gemz\Port\Tests;

use Gemz\Port\Exceptions\InvalidArgument;
use Gemz\Port\Port;
use PHPUnit\Framework\TestCase;

class PortTest extends TestCase
{
    /** @var string */
    protected $domain = 'gemz.io';

    /** @var Port */
    protected $port;

    public function setUp(): void
    {
        $this->port = new Port($this->domain);
    }

    public function test_throws_exception_if_port_not_valid(): void
    {
        $this->expectException(InvalidArgument::class);

        $this->port->check('test');
    }

    public function test_throws_exception_if_protocol_not_valid(): void
    {
        $this->expectException(InvalidArgument::class);

        $this->port->check([80 => 'abc']);
    }

    public function test_throws_exception_if_port_and_protocol_not_valid(): void
    {
        $this->expectException(InvalidArgument::class);

        $this->port->check(['987' => 'abc']);
    }

    public function test_can_check_one_port(): void
    {
        $checks = $this->port
            ->setTimeout(0.1)
            ->check(80);

        $this->assertArrayHasKey('port', $checks[0]);
        $this->assertArrayHasKey('protocol', $checks[0]);
        $this->assertArrayHasKey('open', $checks[0]);
    }

    public function test_can_check_multiple_ports_as_integers(): void
    {
        $checks = $this->port
            ->setTimeout(0.1)
            ->check(80, 443);

        $this->assertArrayHasKey('port', $checks[0]);
        $this->assertArrayHasKey('protocol', $checks[0]);
        $this->assertArrayHasKey('open', $checks[0]);
    }

    public function test_can_check_multiple_ports_with_protocols_as_array(): void
    {
        $checks = $this->port
            ->setTimeout(0.1)
            ->check([80 => 'tcp', 443 => 'tls']);

        $this->assertTrue(count($checks) == 2);
        $this->assertArrayHasKey('port', $checks[0]);
        $this->assertArrayHasKey('protocol', $checks[0]);
        $this->assertArrayHasKey('open', $checks[0]);
    }

    public function test_throws_exception_if_domain_is_empty(): void
    {
        $this->expectException(InvalidArgument::class);

        $checks = Port::for('')
            ->setTimeout(0.1)
            ->check([80 => 'tcp', 443 => 'tls']);
    }

    public function test_can_get_domain(): void
    {
        $domain = $this->port->getDomain();

        $this->assertSame($domain, $this->domain);
    }

    public function test_can_get_default_ports(): void
    {
        $ports = $this->port->getDefaultPorts();

        $this->assertArrayHasKey(80, $ports);
    }

    public function test_can_get_protocols(): void
    {
        $protocols = $this->port->getProtocols();

        $this->assertTrue(in_array('tcp', $protocols));
        $this->assertTrue(in_array('ssl', $protocols));
        $this->assertTrue(in_array('tls', $protocols));
        $this->assertTrue(in_array('udp', $protocols));
    }

    public function test_can_set_tcp_protocol_for_all_ports(): void
    {
        $checks = $this->port->useTcp()->check(80, 443);

        $this->assertTrue($checks[0]['protocol'] == 'tcp');
    }

    public function test_can_set_tls_protocol_for_all_ports(): void
    {
        $checks = $this->port->useTls()->check(80, 443);

        $this->assertTrue($checks[0]['protocol'] == 'tls');
    }

    public function test_can_set_ssl_protocol_for_all_ports(): void
    {
        $checks = $this->port->useSsl()->check(80, 443);

        $this->assertTrue($checks[0]['protocol'] == 'ssl');
    }

    public function test_can_set_udp_protocol_for_all_ports(): void
    {
        $checks = $this->port->useUdp()->check(80, 443);

        $this->assertTrue($checks[0]['protocol'] == 'udp');
    }

    public function test_can_override_protocol_for_specific_ports(): void
    {
        $checks = $this->port->useUdp()->check([80 => 'tcp', 443 => 'tls']);

        $this->assertTrue($checks[0]['protocol'] == 'tcp');
    }

}


src/Port.php
<?php

namespace Gemz\Port;

use React\Promise\Deferred;
use React\Promise\PromiseInterface;
use React\Socket\Connector;
use React\EventLoop\Factory;
use React\EventLoop\LoopInterface;
use React\Socket\ConnectionInterface;
use Gemz\Port\Exceptions\InvalidArgument;

class Port
{
    const PROTOCOL_TCP = 'tcp';
    const PROTOCOL_UDP = 'udp';
    const PROTOCOL_SSL = 'ssl';
    const PROTOCOL_TLS = 'tls';

    /** @var string */
    protected $domain;

    /** @var string */
    protected $protocol = self::PROTOCOL_TCP;

    /** @var float */
    protected $timeout = 0.25;

    /** @var array */
    protected $protocols = ['tcp', 'tls', 'udp', 'ssl'];

    /**
     * @var array
     */
    protected $defaultPorts = [
        22 => 'tcp',
        80 => 'tcp',
        443 => 'tls',
        8080 => 'tcp',
        3306 => 'tcp',
    ];

    /** @var array */
    protected $result = [];

    /** @var array */
    protected $portsWithProtocol = [];

    public static function for(string $domain): self
    {
        return new self($domain);
    }

    public function __construct(string $domain)
    {
        $this->domain = $this->sanitizeDomain($domain);
    }

    protected function sanitizeDomain(string $domain): string
    {
        if (empty($domain)) {
            throw InvalidArgument::domainIsNotValid($domain);
        }

        return str_replace(['http://', 'https://'], '', $domain);
    }

    public function getDomain(): string
    {
        return $this->domain;
    }

    public function getDefaultPorts(): array
    {
        return $this->defaultPorts;
    }

    public function getProtocols(): array
    {
        return $this->protocols;
    }

    protected function addResult(int $port, string $protocol, bool $isOpen): void
    {
        array_push($this->result, [
           'port' => $port,
           'protocol' => $protocol,
           'open' => $isOpen
        ]);
    }

    /**
     * @param array|int ...$ports
     *
     * @return array
     */
    public function check(...$ports): array
    {
        $this->resolvePorts($ports);

        foreach ($this->portsWithProtocol as $port => $protocol) {
            $this->checkPort($port, $protocol)
                ->then(
                    function () use ($port, $protocol) {
                        $this->addResult($port, $protocol, true);
                    },
                    function () use ($port, $protocol) {
                        $this->addResult($port, $protocol, false);
                    }
                );
        }

        return $this->result;
    }

    protected function getUri(string $protocol): string
    {
        return "{$protocol}://{$this->domain}";
    }

    protected function checkPort(int $port, string $protocol): PromiseInterface
    {
        $deferred = new Deferred();

        $portIsOpen = ($protocol == self::PROTOCOL_UDP)
            ? $this->checkUdpPort($port)
            : $portIsOpen = $this->checkNonUdpPort($port, $protocol);

        if ($portIsOpen) {
            $deferred->resolve();
        } else {
            $deferred->reject();
        }

        return $deferred->promise();
    }

    protected function checkNonUdpPort(int $port, string $protocol): bool
    {
        $handler = @fsockopen(
            $this->getUri($protocol), $port, $errno, $errstr, $this->timeout
        );

        if (is_resource($handler)) {
            fclose($handler);
            return true;
        }

        return false;
    }

    protected function checkUdpPort($port): bool
    {
        $handler = @fsockopen(
            $this->getUri(self::PROTOCOL_UDP), $port, $errno, $errstr, $this->timeout
        );

        if (! $handler) {
            return false;
        }

        socket_set_timeout($handler, $this->timeout);

        $write = fwrite($handler, "x00");

        if (! $write) {
            return false;
        }

        $startTime = time();

        fread($handler, 1);

        $endTime = time();

        $timeDiff = $endTime - $startTime;

        fclose($handler);

        if ($timeDiff >= $this->timeout) {
            return true;
        }

        return false;
    }

    public function useTcp(): self
    {
        $this->setProtocol(self::PROTOCOL_TCP);

        return $this;
    }

    public function useTls(): self
    {
        $this->setProtocol(self::PROTOCOL_TLS);

        return $this;
    }

    public function useUdp(): self
    {
        $this->setProtocol(self::PROTOCOL_UDP);

        return $this;
    }

    public function useSsl(): self
    {
        $this->setProtocol(self::PROTOCOL_SSL);

        return $this;
    }

    public function setTimeout(float $seconds): self
    {
        $this->timeout = $seconds;

        return $this;
    }

    protected function setProtocol(string $protocol): void
    {
        $this->protocol = $protocol;
    }

    protected function addPortWithProtocol(int $port, string $protocol): void
    {
        $this->portsWithProtocol[$port] = $protocol;
    }

    protected function resolvePorts(array $ports = []): void
    {
        if (empty($ports)) {
            $ports = $this->defaultPorts;
        }

        $ports = is_array($ports[0] ?? null)
            ? $ports[0]
            : $ports;

        foreach ($ports as $port => $protocol) {

            if (in_array($protocol, $this->protocols) && is_int($port)) {
                $this->addPortWithProtocol($port, $protocol);
                continue;
            }

            if (is_int($protocol)) {
                $this->addPortWithProtocol($protocol, $this->protocol);
                continue;
            }

            throw InvalidArgument::portOrProtocolIsNotValid($port, $protocol);
        }
    }
}
